File: 2018/05/rs/src/main.rs
fn part1(polymer: &str) -> usize {
    let mut polymer_ints = polymer
        .as_bytes()
        .iter()
        .map(|c| *c as i8)
        .collect::<Vec<i8>>();
    let mut had_changes = true;
    while had_changes {
        had_changes = false;
        let mut index = 0;
        while index < polymer_ints.len() - 1 {
            if i8::abs(polymer_ints[index] - polymer_ints[index + 1]) == 32 {
                polymer_ints.remove(index as usize);
                polymer_ints.remove(index as usize);
                had_changes = true;
            } else {
                index += 1;
            }
        }
    }
    polymer_ints.len()
}

fn part2(polymer: &str) -> usize {
    let mut min_units = usize::MAX;
    for c_ord in b'A'..(b'Z' + 1) {
        let stripped_polymer = polymer.replace((c_ord as char).to_string().as_str(), "");
        let stripped_polymer = stripped_polymer.replace(((c_ord + 32) as char).to_string().as_str(), "");
        min_units = min_units.min(part1(&stripped_polymer));
    }
    min_units
}

fn solve(polymer: &str) -> (usize, usize) {
    (part1(polymer), part2(polymer))
}

fn get_input(file_path: &String) -> String {
    String::from(
        std::fs::read_to_string(file_path)
            .expect("Error reading input file!")
            .trim(),
    )
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = std::time::Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:7}", now.elapsed().as_secs_f32());
}

File: 2018/05/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <string.h>
#include <math.h>
#include <limits.h>

typedef char *Input;
typedef struct
{
    int part1;
    int part2;
} Results;

int countNoRepeats(char *polymer, char ignore)
{
    int length = strlen(polymer);
    char *letters = malloc(strlen(polymer) + 1);
    strcpy(letters, polymer);
    int hasChanges = 1, index, lastLetterIndex;
    char letter, lastLetter, upperIgnore = ignore - 32;
    while (hasChanges)
    {
        hasChanges = 0;
        index = 0;
        lastLetter = 0;
        lastLetterIndex = 0;
        while (index < length)
        {
            letter = letters[index++];
            if (letter == 0) continue;
            if (letter == ignore || letter == upperIgnore)
            {
                hasChanges = 1;
                letters[index - 1] = 0;
                continue;
            }
            if (abs(letter - lastLetter) == 32)
            {
                hasChanges = 1;
                lastLetter = 0;
                letters[lastLetterIndex] = letters[index - 1] = 0;
                continue;
            }
            lastLetter = letter;
            lastLetterIndex = index - 1;
        }
    }
    free(letters);
    int count = 0;
    while (length--)
        count += *(letters++) != 0;
    return count - 1;
}

int part2(Input input)
{
    int min = INT_MAX, test;
    for (char c = 'a'; c <= 'z'; c++)
        min = (test = countNoRepeats(input, c)) < min ? test : min;
    return min;
}

Results solve(Input input)
{
    return (Results){countNoRepeats(input, 0), part2(input)};
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    fseek(file, 0, SEEK_END);
    long length = ftell(file);
    rewind(file);
    char *input = malloc(length + 1);
    fread(input, 1, length, file);
    fclose(file);
    return input;
}

void freeInput(Input input)
{
    free(input);
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}